### DISCUSSION TITLE
"Burst Mode" is fast, but the screen now is flickering and finally dead...

### DISCUSSION THREAD

**Discussion by JavierPascual** [RudolphRiedel/FT800-FT813#68] [Q&A]
Hi all!

Congrats to Rudolph for the library, you are a beast! :-D

I'm using the burst mode, and the speed has been increase a lot of... but it's impossible to maintain stable the display.

When I use the "non burst" commands, everything is ok, but with "_burst" the flickering is bizarre.

I'm using the next commands in the main loop:

`void EVE_prepare(uint32_t _bg_color)
{
    EVE_start_cmd_burst();
    EVE_cmd_dl_burst(CMD_DLSTART);
    EVE_cmd_dl_burst(DL_CLEAR_RGB | _bg_color);
    EVE_cmd_dl_burst(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
    EVE_cmd_dl_burst(VERTEX_FORMAT(0));
}

void EVE_display()
{
    EVE_cmd_dl_burst(DL_DISPLAY);
    EVE_cmd_dl_burst(CMD_SWAP);
    EVE_end_cmd_burst();
}`

**Comment by jwestes** [comment]
I'm using burst in my application and I'm not seeing any flicker.

**Comment by RudolphRiedel** [comment]
Flickering usually occurs when the display refresh is faster than the frames per second the display actually is driven with.
Soo, what module exactly and how fast is your refresh?

**Comment by JavierPascual** [comment]
First of all, I had some fatal errors in the SPI bus. I have changed the EVE_start_dma_transfer() function because the spi_device_queue_trans() function is generating problems with multitasking. With spi_device_transmit() and some little other changes (clear CS, etc), everything is ok.

About the flickering, it disappear when I put a MS_Delay(2) in my painting function, but then the performance of my controls & interface is very poor:

`
void EVE_paint() {

    EVE_cmd_dl_burst(DL_DISPLAY);

    EVE_cmd_dl_burst(CMD_SWAP);

    EVE_end_cmd_burst();

    DELAY_MS(2);

}`

Without the MS_Delay(2), the performance is perfect (> 150 fps) but the flickering is BIG and stability is terrible, in only 3-5 seconds the screen (ILI9341) is dead with a slow white fade out.

**Reply by RudolphRiedel** [reply]
Is this ESP32?
The last time I checked, ESP32 and ESP32-S3 was fine - are you using ESP-IDF 5.0 already?

How is spi_device_queue_trans() causing issues with multitasking when all it does is to put in a friendly request for the RTOS to put it into the DMA queue so that it can be eventually picked up by the RTOS DMA handler?
Granted, it takes forever untill something finally is transmitted, but that is an issue with ESP IDF in general.

And "> 150 fps"? As I wrote, you can't refresh (which means issue CMD_SWAP) faster than the display can show the frames.
Usually the timings for the displays are tuned around 60Hz.

**Comment by JavierPascual** [comment]
Using ESP32 with ESP-IDF 4.4.3 + BT817Q + ILI9341.

Ok, I'm going to try again with spi_device_queue_trans(), to check the last error I received.

About FPS... how can I slow down each loop to sync with 60 fps without MS_DELAY()?

Thats my EVE config:

#define EVE_HSIZE (240L) /* Thd Length of visible part of line (in PCLKs) - display width */
#define EVE_HSYNC0 (20L)   /* Thf Horizontal Front Porch */
#define EVE_HSYNC1 (80L)   /* Thf + Thp Horizontal Front Porch plus Hsync Pulse width */
#define EVE_HOFFSET (120L) /* Thf + Thp + Thb Length of non-visible part of line (in PCLK cycles) */
#define EVE_HCYCLE (400L)  /* Th Total length of line (visible and non-visible) (in PCLKs) */
#define EVE_VSIZE (320L)  /* Tvd Number of visible lines (in lines) - display height */
#define EVE_VSYNC0 (3L)    /* Tvf Vertical Front Porch */
#define EVE_VSYNC1 (6L)    /* Tvf + Tvp Vertical Front Porch plus Vsync Pulse width */
#define EVE_VOFFSET (20L)  /* Tvf + Tvp + Tvb Number of non-visible lines (in lines) */
#define EVE_VCYCLE (400L)  /* Tv Total number of lines (visible and non-visible) (in lines) */
#define EVE_ROTATE (6L)    /* Portrait */
#define EVE_BACKLIGHT_PWM (128L) /* 100% */
#define EVE_PCLKPOL (0L)   /* PCLK polarity (0 = rising edge, 1 = falling edge) */
#define EVE_SWIZZLE (0L)   /* Defines the arrangement of the RGB pins */
#define EVE_PARAM_CLKSEL (0x46)  /* 0x46 = 72 Mhz*/
#define EVE_PARAM_FREQUENCY (72000000U) /* 72 Mhz */
#define EVE_PCLK (1L)      /* 1 = use second PLL for pixel-clock in BT817 / BT818 */
#define EVE_PCLK_FREQ (24UL * 1000UL * 1000UL) /* 24 MHz - value for EVE_cmd_pclkfreq */
#define EVE_CSPREAD (0L) /* helps with noise, when set to 1 fewer signals are changed simultaneously, reset-default: 1 */
#define EVE_DITHER (0L)
#define EVE_RENDERMODE (0L)
#define EVE_HAS_CRYSTAL
#define EVE_GEN 4

**Comment by RudolphRiedel** [comment]
Regarding the timing, I have an ESP32 example:
https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_ESP32_PlatformIO

In a nutshell I am using esp_timer_get_time() to execute the main loop every 5ms.

As for the timings, these result in 150 Hz and I wonder if the panel is actually working this fast.
A more apropriate value would be either with EVE_PCLK (7L) to not use the second PLL which gives a pixel clock of about 10MHz and a resulting frame rate of 64,29 Hz.
Or, well, use EVE_PCLK_FREQ (10000000UL) with EVE_PCLK (1L) for 62.5Hz.
And the timings are a bit odd, HCYCLE seems to be a bit high.

>#define EVE_PARAM_CLKSEL (0x46) / 0x46 = 72 Mhz*/
>#define EVE_PARAM_CLKSEL (72000000U) /* 72 Mhz /
>#define EVE_PARAM_CLKSEL (0L)

These are not in my code.
Just saying, remote debugging is difficult enough already when looking at the same code base.
